Src/FluentAssertions/Common/MemberPathSegmentEqualityComparer.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System.Collections.Generic;
+using System.Text.RegularExpressions;
+
+namespace FluentAssertions.Common
+{
+    /// <summary>
+    /// Compares two segments of a <see cref="MemberPath"/>.
+    /// Sets the <see cref="AnyIndexQualifier"/> equal with any numeric index qualifier.
+    /// All other comparisons are default string equality.
+    /// </summary>
+    internal class MemberPathSegmentEqualityComparer : IEqualityComparer<string>
+    {
+        private const string AnyIndexQualifier = "*";
+        private static readonly Regex IndexQualifierRegex = new(@"^\d+$");
+
+        /// <summary>
+        /// Compares two segments of a <see cref="MemberPath"/>.
+        /// </summary>
+        /// <param name="x">Left part of the comparison.</param>
+        /// <param name="y">Right part of the comparison.</param>
+        /// <returns>True if segments are equal, false if not.</returns>
+        public bool Equals(string x, string y)
+        {
+            if (x is null || y is null)
+            {
+                return x == y;
+            }
+
+            if (x == AnyIndexQualifier)
+            {
+                return EqualsAnyIndexQualifier(y);
+            }
+
+            if (y == AnyIndexQualifier)
+            {
+                return EqualsAnyIndexQualifier(x);
+            }
+
+            return x == y;
+        }
+
+        private static bool EqualsAnyIndexQualifier(string segment)
```

Comment:
⛏️ I think naming it `IsIndexQualifier` is a easier to understand.

Reply:
done 👍